core/commands/ls.go

```diff
@@ -60,158 +67,243 @@ The JSON output contains type information.
+				if err != nil {
+					return err
 				}
-				output[i].Links[j] = LsLink{
-					Name: link.Name,
-					Hash: link.Cid.String(),
-					Size: link.Size,
-					Type: t,
+				outputLinks[0] = *lsLink
+				output[i].Links = outputLinks
+				if err = res.Emit(&LsOutput{output}); err != nil {
+					return err
 				}
 			}
 		}
-
-		res.SetOutput(&LsOutput{output})
+		return nil
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+	PostRun: cmds.PostRunMap{
+		cmds.CLI: func(res cmds.Response, re cmds.ResponseEmitter) error {
+			req := res.Request()
+			headers, _ := req.Options[lsHeadersOptionNameTime].(bool)
+			stream, _ := req.Options[lsStreamOptionName].(bool)
 
-			v, err := unwrapOutput(res.Output())
-			if err != nil {
-				return nil, err
+			// in streaming mode we can't automatically align the tabs
+			// so we take a best guess
+			var minTabWidth int
+			if stream {
+				minTabWidth = 10
+			} else {
+				minTabWidth = 1
 			}
 
-			headers, _, _ := res.Request().Option(lsHeadersOptionNameTime).Bool()
-			output, ok := v.(*LsOutput)
-			if !ok {
-				return nil, e.TypeErr(output, v)
-			}
+			multipleFolders := len(req.Arguments) > 1
+			lastDirectoryWritten := -1
 
-			buf := new(bytes.Buffer)
-			w := tabwriter.NewWriter(buf, 1, 2, 1, ' ', 0)
-			for _, object := range output.Objects {
-				if len(output.Objects) > 1 {
-					fmt.Fprintf(w, "%s:\n", object.Hash)
+			tw := tabwriter.NewWriter(os.Stdout, minTabWidth, 2, 1, ' ', 0)
+			for {
+				v, err := res.Next()
+				if err != nil {
+					if err == io.EOF {
+						if multipleFolders {
+							fmt.Fprintln(os.Stdout)
+						}
+						return nil
+					}
+
+					return err
 				}
-				if headers {
-					fmt.Fprintln(w, "Hash\tSize\tName")
+
+				output, ok := v.(*LsOutput)
+				if !ok {
+					return e.TypeErr(output, v)
 				}
-				for _, link := range object.Links {
-					if link.Type == unixfs.TDirectory {
-						link.Name += "/"
+
+				for i, object := range output.Objects {
+					if len(object.Links) == 0 {
+						continue
+					}
+					if i > lastDirectoryWritten {
+						if i > 0 {
+							if multipleFolders {
+								fmt.Fprintln(tw)
+							}
+						}
+						if multipleFolders {
+							fmt.Fprintf(tw, "%s:\n", object.Hash)
+						}
+						if headers {
+							fmt.Fprintln(tw, "Hash\tSize\tName")
+						}
+						lastDirectoryWritten = i
+					}
+					for _, link := range object.Links {
+						if link.Type == unixfs.TDirectory {
+							link.Name += "/"
+						}
+
+						fmt.Fprintf(tw, "%s\t%v\t%s\n", link.Hash, link.Size, link.Name)
 					}
-					fmt.Fprintf(w, "%s\t%v\t%s\n", link.Hash, link.Size, link.Name)
-				}
-				if len(output.Objects) > 1 {
-					fmt.Fprintln(w)
 				}
+				tw.Flush()
 			}
-			w.Flush()
-
-			return buf, nil
 		},
 	},
 	Type: LsOutput{},
 }
+
+func makeDagNodeLinkResults(req *cmds.Request, dagnode ipld.Node) <-chan unixfs.LinkResult {
+	linkResults := make(chan unixfs.LinkResult)
+	go func() {
```

Comment:
So since we need to keep the links in memory for the time of iteration anyways, I think it might be better to just do

```go
links := dagnode.Links()
linkResults := make(chan unixfs.LinkResult, len(links))
for _, l := range dagnode.Links() {
 linkResults <- unixfs.LinkResult{
  Link: l,
  Err:  nil,
 }
}
close(linkResults)
return linkResults

```

Reply:
I agree, but we can clean this up afterwards.